app/helpers/generic_file_helper.rb

```diff
@@ -33,41 +33,34 @@ def render_download_link text = nil
   # download_path(@generic_file, file: 'webm')
   # download_path(id: @asset)
   # download_path document, file: 'thumbnail'
+  #
+  # we shouldn't be overloading this to accept everything under the sun as it leads to all
+  # manner of bugs and corner cases, but these usages are imposed by Hydra/Sufia, so we're forced to live with
+  # it
   def download_path(*args)
-    if args.first.is_a? String
-      item = GenericFile.find(args.first)
-    else
-      item = args.first
-    end
-    return sufia.download_path(*args) unless (item.is_a?(SolrDocument) || item.is_a?(GenericFile))
-    return item.doi_url if item.respond_to?(:doi_url) && item.doi_url.present?
+    raise ArgumentError unless args.present?
+    item = item_for_download(args.shift)
 
-    # not all doi_urls will be in SolrDocuments until a full reindex happens
-    if item.is_a?(SolrDocument) && !item.doi_url_indexed?
-      gf = GenericFile.find(item.id)
-      return gf.doi_url if gf.doi_url.present?
+    # doi supercedes anything else
+    return item.doi_url if item.doi_url.present?
+
+    # this shouldn't happen normally in production, because it indicates an item with no file
```

Comment:
Actually, that will only work on item pages. For download links on search pages, that method isn't called, and won't work, since it assumes it's working with a GenericFile, and it needs to handle SolrDocuments on search and collection pages